src/onehot.jl

```diff
@@ -183,9 +183,23 @@ julia> reshape(1:15, 3, 5) * oh  # this matrix multiplication is done efficientl
  3  6  15  3  9  3  12  3  6  15  3
 ```
 """
-onehotbatch(ls, labels, default...) = _onehotbatch(ls, length(labels) < 32 ? Tuple(labels) : labels, default...)
-# NB function barier:
-_onehotbatch(ls, labels, default...) = batch([onehot(l, labels, default...) for l in ls])
+onehotbatch(data, labels, default...) = _onehotbatch(data, length(labels) < 32 ? Tuple(labels) : labels, default...)
+
+function _onehotbatch(data, labels)
+  indices = UInt32[something(_findval(i, labels), 0) for i in data]
+  if 0 in indices
+    unexpected_values = unique(data[indices .== 0])
+    error("Values $unexpected_values are not in labels")
```

Comment:
What this won't handle is non-indexable data:
```
julia> _onehotbatch((x for x in [10, 20, 10, 20, 30]), [10, 20])
ERROR: MethodError: no method matching getindex(::Base.Generator{Vector{Int64}, typeof(identity)}, ::BitVector)

julia> _onehotbatch("abcd", "abc")
ERROR: ArgumentError: logical indexing not supported for strings
```
Not quite as informative an error, but one compact solution is:
```suggestion
    for x in data
      isnothing(_findval(x, labels)) && error("Value $x not found in labels")
    end
```

Reply:
@mcabbott would this be reasonable? Or overcomplicate things, do you think?

```julia
function _unexpected_values(data, indices)
  unique_values = Set{eltype(data)}()
  it = 1
  for (i, (idx, val)) in enumerate(zip(indices, data))
    it = i
    idx != 0 && continue
    push!(unique_values, val)
    length(unique_values) == 10 && break
  end
  string_values = "(" * join(unique_values, ", ")
  closing_brac = it == length(data) ? ")" : "...)"
  return string_values * closing_brac
end

function _onehotbatch(data, labels)
  indices = UInt32[something(_findval(i, labels), 0) for i in data]
  if 0 in indices
    unexpected_values = _unexpected_values(data, indices)
    error("Values $unexpected_values are not in labels")
  end
  return OneHotArray(indices, length(labels))
end
```

```
julia> onehotbatch((x for x in [10, 20, 10, 20, 30]), [10, 20])
ERROR: Values (30) are not in labels

julia> onehotbatch("abcd", "ab")
ERROR: Values (c, d) are not in labels

julia> onehotbatch(1:10^8, "abcd")
ERROR: Values (5, 4, 6, 7, 2, 10, 9, 8, 3, 1...) are not in labels
```

Reply:
I don't think runtime is a concern at all, but my vote is that this is probably more complexity than is justified. 

You get this error when you mess up and use labels "dog, cat, ..." with data [1,3,2, ...], right? There's some value to printing one wrong point, to lead you to the mistake, but probably not much value in collecting all of them.

But you could equally well mess up the data or the labels. Maybe better than 
```
ERROR: Values (5, 4, 6, 7, 2, 10, 9, 8, 3, 1...) are not in labels
```
would be something like this?
```
ERROR: Value 5 not found in labels "abcd"
```

Reply:
That's fair enough, maybe I'll use your initial suggestion then (just in case labels is a really large array/string).